### PR TITLE
feat: include todo list in summarization output

### PR DESCRIPTION
**Pull request by fkesheh**
## Summary
- Preserve active tasks across conversation summarization by including the current todo list alongside the context summary
- `buildSummaryMessage` now accepts an optional `Todo[]` and appends a `<current_todos>` XML block when todos are non-empty
- `checkAndSummarizeIfNeeded` passes todos through from the call site in `chat-handler.ts` via `getTodoManager().getAllTodos()`

## Test plan
- [x] `npx tsc --noEmit --skipLibCheck` passes
- [x] All 10 summarization tests pass (8 existing + 2 new)
- [x] 100% line coverage on summarization module
- [x] New test: verifies `<current_todos>` block with formatted todo lines when todos exist
- [x] New test: verifies no `<current_todos>` block when todos are empty

## Summary by CodeRabbit

* **New Features**
  * Summaries now include a formatted todo section, retain recent messages for continuity, and support incremental/first-time condensation with improved prompts and token-threshold handling.
  * Summarization can be cancelled via abort signals.

* **Bug Fixes**
  * Summary invalidation is more selective and time-aware; persistence and save failures no longer block UI flow.
  * Message re-truncation ensures token budgets are respected after inserting summaries.

* **Tests**
  * Added comprehensive unit tests covering summarization, todos, aborts, persistence, and edge cases.

### CODE REVIEW

File: convex/messages.ts [hunk 1]
```diff
@@ -39,20 +36,46 @@ const checkAndInvalidateSummary = async (
...
+
+    if (!cutoffMessage) {
       await ctx.db.patch(chat._id, {
         latest_summary_id: undefined,
         update_time: Date.now(),
       });
+      try {
+        await ctx.db.delete(chat.latest_summary_id);
+      } catch {
+        // Summary might already be deleted, ignore
```
Comment (mschead) [inline convex/messages.ts:60]: We need to address this differently because it can lead to a false negative (a summary that exists but, for some other reason, couldn’t be deleted). I searched the convex lib, and they don't have something like `deleteIfExists`. What do you think about implementing it like this?

```javascript
   const summaryDoc = await ctx.db.get(chat.latest_summary_id);
   if (summaryDoc) {
     await ctx.db.delete(chat.latest_summary_id);
   }
```

File: convex/messages.ts [hunk 2]
```diff
@@ -39,20 +36,46 @@ const checkAndInvalidateSummary = async (
...
+
+    if (shouldInvalidate) {
+      await ctx.db.patch(chat._id, {
+        latest_summary_id: undefined,
+        update_time: Date.now(),
+      });
+      try {
+        await ctx.db.delete(chat.latest_summary_id);
+      } catch {
+        // Summary might already be deleted, ignore
```
Comment (mschead) [inline convex/messages.ts:77]: Same thing here.

File: lib/chat/summarization/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  if (!chatId) return;
+
+  try {
+    await saveChatSummary({
+      chatId,
+      summaryText,
+      summaryUpToMessageId: cutoffMessageId,
+    });
+  } catch (error) {
+    console.error("[Summarization] Failed to save summary:", error);
```
Comment (mschead) [inline lib/chat/summarization/helpers.ts:153]: From what I understood, this should rethrow the error. If you check the caller function

```javascript
  const summaryMessage = buildSummaryMessage(summaryText, todos);
  await persistSummary(chatId, summaryText, cutoffMessageId);
  writeSummarizationCompleted(writer);
```

It will continue to execute as if the error didn't happen at all. Is this the expected behavior?

File: lib/chat/summarization/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    parts: [{ type: "text", text }],
+  };
+};
+
+export const persistSummary = async (
+  chatId: string | null,
+  summaryText: string,
+  cutoffMessageId: string,
+): Promise<void> => {
+  if (!chatId) return;
```
Comment (mschead) [inline lib/chat/summarization/helpers.ts:144]: Can we really have a null `chatId`?

File: lib/chat/summarization/__tests__/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+  });
+
+  const getIsolatedModule = () => {
+    let isolatedModule: typeof import("../index");
+
+    jest.isolateModules(() => {
+      jest.doMock("server-only", () => ({}));
+
+      jest.doMock("ai", () => ({
+        generateText: mockGenerateText,
```
Comment (mschead) [inline lib/chat/summarization/__tests__/index.test.ts:35]: I have a suggestion about the tests that could be scoped to another task, but since new tests were added I want to raise it now: we could try to avoid mocks as much as possible, because they tend to make tests brittle. Take this test as an example:

`should summarize and return correct structure when threshold exceeded`

It mocks `token count (9500)` and `max (10000)` instead of using the real threshold. If someone changes `SUMMARIZATION_THRESHOLD_PERCENTAGE`, these tests won’t catch it—they never use the constant, so that behavior isn’t actually tested.

As I said, the idea is not to tackle this now, but just to open the discussion. What do you think?

File: lib/chat/summarization/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            "Provide a technically precise summary of the above conversation segment that preserves all operational security context while keeping the summary concise and to the point.",
+        },
+      ],
+    });
+    return result.text;
+  } catch (error) {
+    if (abortSignal?.aborted) {
+      throw error;
+    }
+    console.error("[Summarization] Failed to generate summary:", error);
```
Comment (mschead) [inline lib/chat/summarization/helpers.ts:114]: Is it ok to skip errors here as well?

File: lib/chat/summarization/helpers.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  if (todos.length > 0) {
+    const todoLines = todos
+      .map((todo) => `- [${todo.status}] ${todo.content}`)
+      .join("\n");
+    text += `\n<current_todos>\n${todoLines}\n</current_todos>`;
+  }
+
+  return {
+    id: uuidv4(),
+    role: "user",
```
Comment (mschead) [inline lib/chat/summarization/helpers.ts:134]: I think we should use a role type for this case.

Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:134]: There is no type in the codebase. Its typed as const enum, not a string

File: lib/chat/summarization/helpers.ts (hunk 4 above)
Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:144]: Yes for temporary chats

File: convex/messages.ts (hunk 2 above)
Comment (fkesheh) [reply convex/messages.ts:77]: I think is 2x the cost and we are silently failing over no matter what is the error

File: lib/chat/summarization/helpers.ts (hunk 6 above)
Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:114]: I think the idea of summarization is silently failing over

File: lib/chat/summarization/helpers.ts (hunk 7 above)
Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:134]: role: 'system' | 'user' | 'assistant';

File: convex/messages.ts (hunk 1 above)
Comment (fkesheh) [reply convex/messages.ts:60]: I think is 2x the cost and we are silently failing over no matter what is the error

Comment (mschead) [reply convex/messages.ts:60]: You're right. What do you think about adding a `console.error`? Just to not have a silenced error.

File: convex/messages.ts (hunk 2 above)
Comment (mschead) [reply convex/messages.ts:77]: Same here as I suggested before.

File: lib/chat/summarization/helpers.ts (hunk 6 above)
Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:114]: This is the current production code, it was just moved.

File: lib/chat/summarization/helpers.ts (hunk 3 above)
Comment (fkesheh) [reply lib/chat/summarization/helpers.ts:153]: This is the current production code, it was just moved.
